pep-0669.rst

```diff
@@ -0,0 +1,318 @@
+
+Checkpoints
+-----------
+
+A checkpoint is simply a point in code defined by a 
+``(codeobject, offset)`` pair.
+Every time a checkpoint is reached, the registered callable is called.
+
+Instrumentation
+---------------
+
+Instrumentation supports the bulk insertion of checkpoints, but does not
+allow insertion or removal of checkpoints after code has started to execute.
+
+The events are::
+
+  * BRANCH: A conditional branch is reached. 
+  * JUMPBACK: A backwards, unconditional branch is reached.
+  * ENTER: A Python function is entered.
+  * EXIT: A Python function exits normally (without an exception).
+  * UNWIND: A Python function exits with an unhandled exception.
+  * C_CALL: A call to any object that is not a Python function.
+  * C_RETURN: A return from any object that is not a Python function.
+  * RAISE: An exception is raised.
+  * EXCEPT: An exception is handled.
+
+For each ``ENTER`` event there will be a corresponding
+``EXIT`` or ``UNWIND`` event.
+For each ``C_CALL`` event there will be a corresponding
+``C_RETURN`` or ``RAISE`` event.
+
+All events are integer powers of two and can be bitwise or-ed together to
+instrument multiple events.
+
+Instrumenting code objects
+''''''''''''''''''''''''''
+
+Code objects can be instrumented by calling::
+
+  instrumentation.instrument(codeobject, events)
+
+Code objects must be instrumented before they are executed.
+An exception will be raised if the code object has been executed before it
+is instrumented.
+
+Register callback functions for instrumentation
+'''''''''''''''''''''''''''''''''''''''''''''''
+
+To register a callable for events call::
+
+  instrumentation.register(event, func)
+
+Functions can be unregistered by calling
+``instrumentation.register(event, None)``.
+
+Callback functions can be registered at any time.
+
+Callback function arguments
+'''''''''''''''''''''''''''
+
+When an event occurs the registered function will be called.
+The arguments provided are as follows:
+
+* BRANCH: ``func(code: CodeType, offset: int, taken:bool)``
+* JUMPBACK: ``func(code: CodeType, offset: int)``
+* ENTER: ``func(code: CodeType, offset: int)``
+* EXIT: ``func(code: CodeType, offset: int)``
+* C_CALL: ``func(code: CodeType, offset: int, value: object)``
+* C_RETURN: ``func(code: CodeType, offset: int, value: object)``
+* C_EXCEPT: ``func(code: CodeType, offset: int, exception: BaseException)``
+* RAISE: ``func(code: CodeType, offset: int, exception: BaseException)``
+* EXCEPT: ``func(code: CodeType, offset: int)``
+* UNWIND: ``func(code: CodeType)``
+
+Monitoring
+----------
+
+Monitoring allows checkpoints to be inserted or removed at any
+point in the program's execution.
+
+The following functions are provided to insert monitoring points::
+
+  instrumentation.insert_monitors(codeobject, *offsets)
+  instrumentation.remove_monitors(codeobject, *offsets)
+  instrumentation.monitor_off(codeobject, offset)
+  instrumentation.monitor_on(codeobject, offset)
```

Comment:
Is `monitor_off(co, offs)` equivalent to `insert_monitors(co, [offs])`? If so, why bother with the single-offset API at all? Certainly a user of this API can be trusted to listify the argument?

Reply:
> Is monitor_off(co, offs) equivalent to insert_monitors(co, [offs])?

No. Turning a monitor on or off is a very cheap (<1us) operation. Inserting or removing a monitor is a very expensive operation maybe taking 100s of milliseconds as it may cause cascading de-optimizations.

So, we want to allow multiple monitoring checkpoints to be inserted at once because:

1. Each insertion (or removal) or monitors can be very expensive.
2. Debuggers need to be able to insert multiple check points for lines that have several instructions (e.g. code in finally blocks, and duplicated tests in while loops).

If a user can't listify the arguments, I don't have much hope of them implementing a debugger :slightly_smiling_face:

Reply:
Oh. This important distinction might be easier to follow if you allowed yourself some mention of the (roughly) intended implementation. In my imagination, inserting monitors is roughly equivalent to

1. Specialize the code object if not already specialized
2. For each offset, replace the specialized opcode with a MONITOR opcode
3. Maybe clear the inline cache for that code object?

But then monitor_off() would seem to be equivalent to replacing the MONITOR opcode with the original opcode (or perhaps the specialized variant, if we care).

hat kind of cascading de-optimizations am I missing? Anything that's currently implemented, or is this referring to e.g. generating machine code?